ImageRef is mutated with the wrong digest

**Describe the bug**

This issue may be precipitated by https://github.com/sse-secure-systems/connaisseur/issues/582. Our CI platform routinely deploys about 40 pods into a fresh namespace to test code changes. We noticed that after deploying connaisseur, some of the time, the pods in these namespaces fail with an error like this:

```
  Warning  Failed       12s                kubelet            Failed to pull image "docker.io/gergorg/rtdb@sha256:8f76cb854b018213597d910fb4a1ad7d5d5ed775fbe328594b9d6347a48c38f9": rpc error: code = NotFound
 desc = failed to pull and unpack image "docker.io/myorg
/rtdb@sha256:8f76cb854b018213597d910fb4a1ad7d5d5ed775fbe328594b9d6347a48c38f9": failed to resolve reference 
"docker.io/myorg/rtdb@sha256:8f76cb854b018213597d910fb4a1ad7d5d5ed775fbe328594b9d6347a48c38f9": docker.io
/myorg/rtdb@sha256:8f76cb854b018213597d910fb4a1ad7d5d5ed775fbe328594b9d6347a48c38f9: not found
```

The digest ref is the result of a deployment being mutated by connaisseur, having the image tag changed into the digest format. 

**However**... the actual digest applied belongs to a **different** image in a **different** repository, which was deployed by CI at the same time into the namespace:

<img width="329" alt="Harbor 2022-03-17 16-26-56" src="https://user-images.githubusercontent.com/1524686/158740945-b2971825-7ec2-41e2-b2d7-95ed7f204cf2.png">

It would seem that connaisseur "mixed up" the digests of the various deployments/pods which were deployed at the same time (potentially due to an error parsing the CRs), and substituted the digest for one image, with another.

**Expected behavior**

Connaisseur only mutates imageRefs matching the repository and image which was originally intercepted

**Optional: To reproduce**

This is hard, since the issue seems to be intermittent. It **seems** to co-inside with the errors generated by https://github.com/sse-secure-systems/connaisseur/issues/582. To reproduce, try deploying about 20 different images simultaneously, mixed in with some CRDs which include an image ref.

**Optional: Versions (please complete the following information as relevant):**
 - Connaisseur: 2.5.1

here, is the exact case where the digest confusion happens (caused digest confusion by deploying several pods with images 'fyll' being unsigned and 'eve' being signed):
```bash
[2022-03-20 15:11:00,820] INFO: COSIGN output of trust root 'default' for image'ghcr.io/xopham/eve:signed': RETURNCODE: 0; STDOUT: {"critical":{"identity":{"docker-reference":"ghcr.io/xopham/eve"},"image":{"docker-manifest-digest":"sha256:8702f190b3b8d06e1976ecbd4b9dbd3e3f7431f2e11baad3a1368ce2a0523586"},"type":"cosign container image signature"},"optional":null}
[2022-03-20 15:11:00,821] INFO: COSIGN output of trust root 'default' for image'ghcr.io/xopham/fyll:signed': RETURNCODE: 1; STDOUT: ; STDERR: Error: no matching signatures:
[2022-03-20 15:11:00,822] INFO: {'message': 'No trust data for image "ghcr.io/xopham/fyll:signed".', 'context': {'trust_data_type': 'dev.cosignproject.cosign/signature', 'stderr': 'Error: no matching signatures:\n\nmain.go:46: error during command execution: no matching signatures:\n', 'image': 'ghcr.io/xopham/fyll:signed', 'trust_root': 'default', 'detection_mode': False}}
[2022-03-20 15:11:00,822] INFO: {'message': 'successful verification of image "ghcr.io/xopham/eve:signed"', 'context': {'user': 'kubernetes-admin', 'operation': 'CREATE', 'kind': 'Pod', 'name': 'single-pod-4', 'namespace': 'default', 'image': 'ghcr.io/xopham/eve:signed'}}
[2022-03-20 15:11:00,823] INFO: {'message': 'successful verification of image "ghcr.io/xopham/fyll:signed"', 'context': {'user': 'kubernetes-admin', 'operation': 'CREATE', 'kind': 'Pod', 'name': 'single-pod-4', 'namespace': 'default', 'image': 'ghcr.io/xopham/fyll:signed'}}
[2022-03-20 15:11:00,916] INFO: 172.18.0.5 - - [20/Mar/2022 15:11:00] "POST /mutate?timeout=30s HTTP/1.1" 200 -
```
Now, unfortunately #541 is still open which makes it hard to derive causality, but at least from time-correlation it seems that `cosign` correctly highlights lack of a valid signature ("*no matching signatures*") for *fyll*. However, the same time a valid signature is found for *eve* (`sha256:8702f190b3b8d06e1976ecbd4b9dbd3e3f7431f2e11baad3a1368ce2a0523586`) and that same instant Connaisseur reports successful verification for *fyll*. It is not clear whether the prior log at `[2022-03-20 15:11:00,822]` is simply overwritten or unrelated. Still, from the Pod's (that should have never even passed) logs, we learn that indeed *fyll* was passed with *eve*'s digest:

```bash
  Normal   Pulling    18m (x3 over 18m)     kubelet            Pulling image "ghcr.io/xopham/fyll@sha256:8702f190b3b8d06e1976ecbd4b9dbd3e3f7431f2e11baad3a1368ce2a0523586"
  Warning  Failed     18m (x3 over 18m)     kubelet            Failed to pull image "ghcr.io/xopham/fyll@sha256:8702f190b3b8d06e1976ecbd4b9dbd3e3f7431f2e11baad3a1368ce2a0523586": rpc error: code = NotFound desc = failed to pull and unpack image "ghcr.io/xopham/fyll@sha256:8702f190b3b8d06e1976ecbd4b9dbd3e3f7431f2e11baad3a1368ce2a0523586": failed to resolve reference "ghcr.io/xopham/fyll@sha256:8702f190b3b8d06e1976ecbd4b9dbd3e3f7431f2e11baad3a1368ce2a0523586": ghcr.io/xopham/fyll@sha256:8702f190b3b8d06e1976ecbd4b9dbd3e3f7431f2e11baad3a1368ce2a0523586: not found
  Warning  Failed     18m (x3 over 18m)     kubelet            Error: ErrImagePull
  Warning  Failed     17m (x4 over 18m)     kubelet            Error: ImagePullBackOff
  Normal   BackOff    3m48s (x65 over 18m)  kubelet            Back-off pulling image "ghcr.io/xopham/fyll@sha256:8702f190b3b8d06e1976ecbd4b9dbd3e3f7431f2e11baad3a1368ce2a0523586"
```
This means that the error does not reside in the subprocess call to `cosign`, but must somehow happen in the processing afterwards, potentially due to parallelism via *concurrent futures* / *asyncio* or the construct of [updating a class variable in the parallel processing loop](https://github.com/sse-secure-systems/connaisseur/blob/master/connaisseur/validators/cosign/cosign_validator.py#L115). Still, technically both validations should happen in [separate class instances of the same validator](https://github.com/sse-secure-systems/connaisseur/blob/master/connaisseur/flask_application.py#L131).
This is very strange behavior.

--- a/connaisseur/validators/cosign/cosign_validator.py
+++ b/connaisseur/validators/cosign/cosign_validator.py
@@ -24,13 +24,11 @@ from connaisseur.validators.interface import ValidatorInterface
 class CosignValidator(ValidatorInterface):
     name: str
     trust_roots: list
-    vals: dict  # validations, a dict for each required trust root containing validated digests or errors
     k8s_keychain: bool
 
     def __init__(self, name: str, trust_roots: list, auth: dict = None, **kwargs):
         super().__init__(name, **kwargs)
         self.trust_roots = trust_roots
-        self.vals = {}
         self.k8s_keychain = False if auth is None else auth.get("k8s_keychain", False)
 
     async def validate(
@@ -43,25 +41,26 @@ class CosignValidator(ValidatorInterface):
             "threshold",
             1 if trust_root != "*" or any(required) else len(self.trust_roots),
         )
-
-        self.vals = self.__get_pinned_keys(trust_root, required, threshold)
+        # vals is a validations dict for each required trust root containing validated digests and errors
+        vals = self.__get_pinned_keys(trust_root, required, threshold)
 
         # use concurrent.futures for now
         # tasks = [self.__validation_task(k, str(image)) for k in self.vals.keys()]
         # await asyncio.gather(*tasks)
 
         # prepare executor
-        num_workers = len(self.vals)
+        num_workers = len(vals)
         executor = ThreadPoolExecutor(num_workers)
         # prepare tasks
-        arguments = [(k, str(image)) for k in self.vals.keys()]
+        # a copy of vals dictionaries is passed to concurrent validation to ensure thread-safe execution
+        arguments = [(k, v.copy(), str(image)) for k, v in vals.items()]
         futures = [executor.submit(self.__validation_task, *arg) for arg in arguments]
-        # await results (output dropped as `self.vals` is updated within function)
+        # await results (output dropped as `vals` is updated within function)
         for future in futures:
-            future.result()
+            vals.update(future.result())
 
         return CosignValidator.__apply_policy(
-            vals=self.vals, threshold=threshold, required=required
+            vals=vals, threshold=threshold, required=required
         )
 
     def __get_pinned_keys(self, key_name: str, required: list, threshold: int):
@@ -105,19 +104,19 @@ class CosignValidator(ValidatorInterface):
         return keys
 
     # async def __validation_task(self, trust_root: str, image: str):
-    def __validation_task(self, trust_root: str, image: str):
+    def __validation_task(self, trust_root: str, val: dict, image: str):
         """
         Async task for each validation to gather all required validations,
         execute concurrently and update results.
         """
         try:
             # self.vals[trust_root]["digest"] = await self.__get_cosign_validated_digests(
-            self.vals[trust_root]["digest"] = self.__get_cosign_validated_digests(
-                image, self.vals[trust_root]
-            )
+            val["digest"] = self.__get_cosign_validated_digests(image, val)
         except Exception as err:
-            self.vals[trust_root]["error"] = err
+            val["error"] = err
             logging.info(err)
+
+        return {trust_root: val}
 
     # async def __get_cosign_validated_digests(self, image: str, trust_root: dict):
     def __get_cosign_validated_digests(self, image: str, trust_root: dict):

--- a/helm/Chart.yaml
+++ b/helm/Chart.yaml
@@ -2,8 +2,8 @@ apiVersion: v2
 name: connaisseur
 description: Helm chart for Connaisseur - a Kubernetes admission controller to integrate container image signature verification and trust pinning into a cluster.
 type: application
-version: 1.3.1
-appVersion: 2.5.1
+version: 1.3.2
+appVersion: 2.5.2
 keywords:
   - container image
   - signature

--- a/helm/values.yaml
+++ b/helm/values.yaml
@@ -1,7 +1,7 @@
 # configure Connaisseur deployment
 deployment:
   replicasCount: 3
-  image: securesystemsengineering/connaisseur:v2.5.1
+  image: securesystemsengineering/connaisseur:v2.5.2
   imagePullPolicy: IfNotPresent
   # imagePullSecrets contains an optional list of Kubernetes Secrets, in Connaisseur namespace,
   # that are needed to access the registry containing Connaisseur image.

--- a/tests/validators/cosign/test_cosign_validator.py
+++ b/tests/validators/cosign/test_cosign_validator.py
@@ -146,7 +146,6 @@ def test_init(index: int, kchain: bool):
     assert val.name == static_cosigns[index]["name"]
     assert val.trust_roots == static_cosigns[index]["trust_roots"]
     assert val.k8s_keychain == kchain
-    assert val.vals == {}
 
 
 @pytest.mark.parametrize(

